geomstats/geometry/waldspace.py

```diff
@@ -0,0 +1,437 @@
+            # TODO element wise minimum of arrays, minimum needed. gs.minimum not impl.
+            x = gs.minimum(gs.maximum(btol, x), 1 - btol)
+            # create the wald
+            return Wald(n=self.n, st=st, x=x)
+
+        # generate samples:
+        prob = prob ** (1 / (self.n - 1))
+        if n_samples == 1:
+            sample = _generate_wald(_prob=prob)
+            print(sample)
+            print(sample.x.shape)
+            return sample.corr
+        return gs.array([_generate_wald(_prob=prob).corr for _ in range(n_samples)])
+
+
+def equivalence_partition(group, relation):
+    """Partitions a set of objects into equivalence classes
+    Taken from
+    'https://stackoverflow.com/questions/38924421/is-there-a-standard-way-to-partition-a
+    n-interable-into-equivalence-classes-given'
+
+    Args:
+        group: collection of objects to be partitioned
+        relation: equivalence relation. I.e. relation(o1,o2) evaluates to True
+            if and only if o1 and o2 are equivalent
+
+    Returns: classes, partitions
+        classes: A sequence of sets. Each one is an equivalence class
+        partitions: A dictionary mapping objects to equivalence classes
+    """
+    classes = []
+    partitions = {}
+    for o in group:  # for each object
+        # find the class it is in
+        found = False
+        for c in classes:
+            if relation(next(iter(c)), o):  # is it equivalent to this class?
+                c.add(o)
+                partitions[o] = c
+                found = True
+                break
+        if not found:  # it is in a new class
+            classes.append({o})
+            partitions[o] = classes[-1]
+    classes = tuple(map(tuple, classes))
+    partitions = {key: tuple(item) for key, item in partitions.items()}
+    return classes, partitions
+
+
+def compute_structure_from_dist(dist, btol=10 ** -10):
+    """Computes the structure induced by a distance matrix, entries = inf possible."""
+    _n = dist.shape[0]
+    _partition, _ = equivalence_partition(group=range(_n),
+                                          relation=lambda i, j: dist[i, j] < math.inf)
+    # for each component, compute the splits in that TREE
+    _split_collection = [
+        _compute_splits_from_sub_dist(sub_dist=dist, sub_labels=labels, btol=btol)
+        for labels in _partition]
+    return Structure(n=_n, partition=_partition, split_sets=_split_collection)
+
+
+def _compute_splits_from_sub_dist(sub_dist, sub_labels, btol=10 ** -10):
+    """Computes the split representation of a single tree characterized by a distance
+    matrix.
+
+    The label set might be smaller than the dimension of dist, in which case,
+    canonically, the sub-matrix is taken.
+    """
+    _n = sub_dist.shape[0]
+    # test if we have only one label, then we have zero splits. note that sub_dist might
+    # be bigger!
+    if len(sub_labels) == 1:
+        return ()
+
+    # plan is the following:
```

Comment:
maybe this longer comments could be insert in the docstring at the beginning of the class